File: src/test/java/no/fintlabs/repository/utils/instanceflowsummary/ParameterizedInstanceFlowTestCaseGenerator.java
package no.fintlabs.repository.utils.instanceflowsummary;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import no.fintlabs.repository.filters.InstanceFlowSummariesQueryFilter;
import no.fintlabs.repository.projections.InstanceFlowSummaryProjection;
import org.testcontainers.shaded.com.google.common.collect.Lists;
import org.testcontainers.shaded.com.google.common.collect.Sets;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.stream.Stream;

import static no.fintlabs.repository.utils.instanceflowsummary.Dataset.ALL_EVENTS_AND_EXPECTED_SUMMARIES;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ParameterizedInstanceFlowTestCaseGenerator {

    public static <T> FilterPropertyTestCases createFilterPropertyTestCases(
            BiConsumer<InstanceFlowSummariesQueryFilter.InstanceFlowSummariesQueryFilterBuilder, T> filterModifier,
            List<FilterPropertyTestCaseConfiguration<T>> individualTestCaseConfigurations,
            List<FilterPropertyTestCaseConfiguration<T>> cartesianTestCaseConfigurations
    ) {
        List<FilterPropertyTestCase> individualTestCases = new ArrayList<>();
        individualTestCaseConfigurations
                .stream()
                .map(testCaseConfiguration -> new FilterPropertyTestCase(
                        filterBuilder -> filterModifier.accept(filterBuilder, testCaseConfiguration.getFilterValue()),
                        testCaseConfiguration.getExpectedSummaries()
                )).forEach(individualTestCases::add);

        List<FilterPropertyTestCase> cartesianTestCases = new ArrayList<>();
        cartesianTestCaseConfigurations
                .stream()
                .map(
                        testCaseConfiguration -> new FilterPropertyTestCase(
                                filterBuilder -> filterModifier.accept(filterBuilder, testCaseConfiguration.getFilterValue()),
                                testCaseConfiguration.getExpectedSummaries()
                        )
                )
                .forEach(cartesianTestCases::add);

        return FilterPropertyTestCases
                .builder()
                .individualTestCaseConfigurations(individualTestCases)
                .cartesianTestCaseConfigurations(cartesianTestCases)
                .build();
    }

    public static List<InstanceFlowSummariesTestCase> combineFilterPropertyTestCases(
            List<FilterPropertyTestCases> filterPropertyTestCases
    ) {
        List<FilterPropertyTestCase> individualTestCases = new ArrayList<>();
        List<List<FilterPropertyTestCase>> cartesianTestCases = new ArrayList<>();

        filterPropertyTestCases.forEach(testCases -> {
            individualTestCases.addAll(testCases.getIndividualTestCaseConfigurations());
            cartesianTestCases.add(testCases.getCartesianTestCaseConfigurations());
        });
        return Stream.concat(
                individualTestCases
                        .stream()
                        .map(List::of)
                        .map(ParameterizedInstanceFlowTestCaseGenerator::createInstanceFlowSummariesTestCase),
                combineAndMapToInstanceFlowSummariesTestCases(cartesianTestCases).stream()
        ).toList();
    }

    private static List<InstanceFlowSummariesTestCase> combineAndMapToInstanceFlowSummariesTestCases(
            List<List<FilterPropertyTestCase>> filterModifierAndPredicateCombinations
    ) {
        return Lists.cartesianProduct(
                        filterModifierAndPredicateCombinations
                                .stream()
                                .peek(l -> l.add(0, new FilterPropertyTestCase(
                                        filterBuilder -> {
                                        },
                                        ALL_EVENTS_AND_EXPECTED_SUMMARIES
                                )))
                                .toList()
                )
                .stream()
                .map(ParameterizedInstanceFlowTestCaseGenerator::createInstanceFlowSummariesTestCase)
                .toList();
    }

    private static InstanceFlowSummariesTestCase createInstanceFlowSummariesTestCase(
            List<FilterPropertyTestCase> filterPropertyTestCases
    ) {
        InstanceFlowSummariesQueryFilter.InstanceFlowSummariesQueryFilterBuilder filterBuilder =
                InstanceFlowSummariesQueryFilter.builder();
        filterPropertyTestCases.forEach(
                filterPropertyTestCase ->
                        filterPropertyTestCase.getFilterModifier().accept(filterBuilder)
        );
        return InstanceFlowSummariesTestCase
                .builder()
                .filter(filterBuilder.build())
                .expectedInstanceFlowSummaries(
                        filterPropertyTestCases
                                .stream()
                                .map(FilterPropertyTestCase::getExpectedSummaries)
                                .reduce(Sets::intersection)
                                .orElse(Set.of())
                                .stream()
                                .map(EventEntitiesAndExpectedSummary::getExpectedSummary)
                                .sorted(Comparator.comparing(InstanceFlowSummaryProjection::getLatestUpdate).reversed())
                                .toList()
                )
                .build();
    }

}

File: src/test/kotlin/no/novari/flyt/history/repository/utils/instanceflowsummary/ParameterizedInstanceFlowTestCaseGenerator.kt
package no.novari.flyt.history.repository.utils.instanceflowsummary

import no.novari.flyt.history.repository.filters.InstanceFlowSummariesQueryFilter
import no.novari.flyt.history.repository.projections.InstanceFlowSummaryProjection

class ParameterizedInstanceFlowTestCaseGenerator private constructor() {
    companion object {
        @JvmStatic
        fun <T> createFilterPropertyTestCases(
            filterModifier: (InstanceFlowSummariesQueryFilter.InstanceFlowSummariesQueryFilterBuilder, T) -> Unit,
            individualTestCaseConfigurations: List<FilterPropertyTestCaseConfiguration<T>>,
            cartesianTestCaseConfigurations: List<FilterPropertyTestCaseConfiguration<T>>,
        ): FilterPropertyTestCases {
            val individualTestCases =
                individualTestCaseConfigurations.map { testCaseConfiguration ->
                    FilterPropertyTestCase(
                        { filterBuilder ->
                            filterModifier(filterBuilder, testCaseConfiguration.filterValue)
                        },
                        testCaseConfiguration.expectedSummaries,
                    )
                }

            val cartesianTestCases =
                cartesianTestCaseConfigurations.map { testCaseConfiguration ->
                    FilterPropertyTestCase(
                        { filterBuilder ->
                            filterModifier(filterBuilder, testCaseConfiguration.filterValue)
                        },
                        testCaseConfiguration.expectedSummaries,
                    )
                }

            return FilterPropertyTestCases
                .builder()
                .individualTestCaseConfigurations(individualTestCases)
                .cartesianTestCaseConfigurations(cartesianTestCases)
                .build()
        }

        @JvmStatic
        fun combineFilterPropertyTestCases(
            filterPropertyTestCases: List<FilterPropertyTestCases>,
        ): List<InstanceFlowSummariesTestCase> {
            val individualTestCases = mutableListOf<FilterPropertyTestCase>()
            val cartesianTestCases = mutableListOf<List<FilterPropertyTestCase>>()

            filterPropertyTestCases.forEach { testCases ->
                individualTestCases += testCases.individualTestCaseConfigurations
                cartesianTestCases += testCases.cartesianTestCaseConfigurations
            }

            return individualTestCases
                .map(::listOf)
                .map(::createInstanceFlowSummariesTestCase) +
                combineAndMapToInstanceFlowSummariesTestCases(cartesianTestCases)
        }

        private fun combineAndMapToInstanceFlowSummariesTestCases(
            filterModifierAndPredicateCombinations: List<List<FilterPropertyTestCase>>,
        ): List<InstanceFlowSummariesTestCase> {
            val lists =
                filterModifierAndPredicateCombinations.map { testCases ->
                    listOf(
                        FilterPropertyTestCase(
                            {},
                            Dataset.ALL_EVENTS_AND_EXPECTED_SUMMARIES,
                        ),
                    ) + testCases
                }

            return cartesianProduct(lists).map(::createInstanceFlowSummariesTestCase)
        }

        private fun createInstanceFlowSummariesTestCase(
            filterPropertyTestCases: List<FilterPropertyTestCase>,
        ): InstanceFlowSummariesTestCase {
            val filterBuilder = InstanceFlowSummariesQueryFilter.builder()
            filterPropertyTestCases.forEach { filterPropertyTestCase ->
                filterPropertyTestCase.filterModifier(filterBuilder)
            }

            return InstanceFlowSummariesTestCase
                .builder()
                .filter(filterBuilder.build())
                .expectedInstanceFlowSummaries(
                    filterPropertyTestCases
                        .map(FilterPropertyTestCase::expectedSummaries)
                        .reduceOrNull(::intersectAll)
                        .orEmpty()
                        .map(EventEntitiesAndExpectedSummary::expectedSummary)
                        .sortedByDescending(InstanceFlowSummaryProjection::latestUpdate),
                ).build()
        }

        private fun intersectAll(
            left: Set<EventEntitiesAndExpectedSummary>,
            right: Set<EventEntitiesAndExpectedSummary>,
        ): Set<EventEntitiesAndExpectedSummary> = left.intersect(right)

        private fun <T> cartesianProduct(lists: List<List<T>>): List<List<T>> {
            return lists.fold(listOf(emptyList())) { acc, list ->
                acc.flatMap { prefix -> list.map(prefix::plus) }
            }
        }
    }
}
